pkg/dynamicconfig/cell.go

```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dynamicconfig
+
+import (
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/statedb"
+	"github.com/spf13/pflag"
+)
+
+var Cell = cell.Module(
+	"cilium-agent-dynamic-config",
+	"Cilium Agent dynamic config map controller",
+	cell.ProvidePrivate(
+		NewConfigTable,
+		NewConfigMapReflector,
+	),
+	cell.Provide(
+		statedb.RWTable[DynamicConfig].ToTable,
+	),
+	cell.Invoke(
+		RegisterConfigMapReflector,
+	),
+	cell.Config(defaultConfig),
+)
+
+var defaultConfig = config{
+	EnableDynamicConfig:  false,
+	DynamicConfigMapName: "cilium-config",
+}
+
+type config struct {
+	EnableDynamicConfig  bool
+	DynamicConfigMapName string
+}
+
+func (c config) Flags(flags *pflag.FlagSet) {
+	flags.Bool("enable-dynamic-config", c.EnableDynamicConfig, "Enables support for dynamic agent config")
```

Comment:
SGTM. I do wonder about how we make the user experience around this clear. I'm thinking for example, maybe a user finds this configuration option, enables it. But then in practice none of the options can be dynamically modified until we start to incrementally introduce support for those things.

I know that on the CFP we had some discussion on this, and there's a tradeoff - I was arguing that there should still be a single configmap so that users don't have to first decide where to configure the options and then figure out the right configmap/resource to configure. On the other hand, when restart is required to modify a setting, it would be nice if the user doesn't get caught out by this, but rather Cilium "just works" to reconfigure the options.

Perhaps a solution to this is to integrate better with the Helm option `rollOutCiliumPods`, so that when the user specifies this option, that enables dynamic config update as well. If Cilium determines that restart is required, it would restart the agents. If restart is not required, ideally the same option would not restart the agents. So then the user just configures one high level option to enable dynamic config, and the details about how the dynamic config gets applied is up to Cilium. Cilium just figures out the most efficient way to apply the config based on the implementation. If this proposal isn't compatible with the implementation details of the rolloutpods setting then it could alternatively be a new Helm flag for dynamic config, we could explore that.

Related, we might consider setting this flag to hidden to reduce the likelihood that a user discovers this flag and then misunderstands the setting, sets it, then is surprised that by itself it doesn't make all the settings dynamic. Another mitigation might be something like the following (would also require a docs update for dynamic configs to close the loop):

```suggestion
	flags.Bool("enable-dynamic-config", c.EnableDynamicConfig, "Enables support for dynamic agent config for supported configurations (see https://docs.cilium.io/en/stable/configuration/dynamic-config/ for a full list)")
```

Reply:
Indeed, totally agree with you. So I think it make sense to hide this for now. I was planning to push these changes so we can actually have some foundation that we can build on and let other devs know that they can use this in their features. 

I was planning to find a way to document this for users, I can add the page for the supported configuration on my task list and document such a page. 

I think we can extend this to support a feature where we let cilium decide if it needs restart or not along with the DynamicLifecycle. So I think then we'll have like 3 options 1. dynamic apply the config changes, 2. FeatureManager could restart(or just start/stop) only specific cells, 3. restart the whole cilium agent.

Reply:
> Perhaps a solution to this is to integrate better with the Helm option rollOutCiliumPods,

@joestringer I suspect it will be difficult to do this, as the way `rollOutCiliumPods` works is by hashing the entire rendered configmap and creating checksum annotation on the pod that will change when the configMap changes so the pod template changes when the configmap does, causing the daemonset to roll out the pods.

To make it integrate in the way your suggesting, will require manually constructing the input values for the checksum calculation, based on whether or not that value is dynamically reloadable. Not to mention, I suspect this behavior would be somewhat confusing. From a devops perspective, I might prefer to have it restart the pods, even if the changed options are dynamically reloadedable when I use `rollOutCiliumPods`. I think they're similar, but distinct features.

Reply:
Just an idea, perhaps in the future we could set a health status to degraded, or emit a specific warning on a Agent pod that detects it needs a restart to have up to date config?  If enabled, would provide a way for users to be notified about whether they have out of sync Agents.

Reply:
Yes, it's on my radar to cover this part mentioned by Chris, I'll make sure to add both as reviewers.